Added screenshots and more details about configuration. (#1130)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/flowref-about.adoc

include::_attributes.adoc[]
endif::[]
:keywords: email, connector, send, retrieve, manage, match, matcher, smtp, pop3, imap

//Anypoint Studio, Design Center connector
[[short_description]]

> Flow Reference routes the Mule event to another flow or sub-flow, and then routes it back within the same Mule application. This process enables you to treat the entire referenced flow like a single component in your current flow.

Reviewer: Can I add this to our Term Presentation list as "the Flow Reference component, or Flow Reference"?

Reviewer: I'm a little confused by "routes to another flow or subflow and then routes it back within the same Mule application."
What is the event routed back through: the original event, the other flow, or the subflow?
And how does this sentence illustrate treating the "entire referenced flow" as a single component in the "current flow"?  Has this topic set previously defined reference flow, another flow, subflow, and current flow?
[If your audience understands this, then never mind my question. I am not as technical as they are.]
I wonder if we should include a diagram?

Reviewer: Also, I suggest "subflow," even though Grammarly doesn't like it, per our style guide entry for prefixes:

prefix hyphenation
The following prefixes don't normally require a hyphen: 

ante, anti, bi, bio, co, counter, extra, infra, inter, intra, macro, mid, mini, multi, 
	non, omni, over, post (after), pre, pro, pseudo, re, semi, sub, super, supra, 
	trans, ultra, un, under.

> Flow Reference breaks up the Mule application into discrete and reusable units. For example, a flow that lists files on a regular basis might reference another flow that processes the output of the List operation. Instead of appending all the processing steps, you can append a Flow Reference that points to the processing flow.

Reviewer: breaks up > breaks (localization)
```suggestion
Flow Reference breaks the Mule application into discrete and reusable units. For example, a flow that lists files on a regular basis might reference another flow that processes the output of the List operation. Instead of appending all the processing steps, you can append a Flow Reference that points to the processing flow.
```
Instead of appending all the processing steps to what? append a Flow Reference to what?

> The next diagram shows the order of processing when one flow references another. When the main flow is triggered, the Mule event travels through and executes the flow until the event reaches the Flow Reference. The event then travels through and executes the referenced flow from beginning to end, and then returns to the main flow.

Reviewer: ```suggestion
The following diagram shows the order of processing when one flow references another:

[consider inserting image here instead of at the end of the current para]

When the main flow is triggered, the Mule event travels through and executes the flow until the event reaches the Flow Reference. The event then travels through and executes the referenced flow from beginning to end, and then returns to the main flow.
```
Ah! This seems to answer at least some of my questions above. Should this explanation come earlier?

image::flowref-about-8b5d1.png[]

== Configure a Flow Reference

To configure a Flow Reference component, follow these steps in Anypoint Studio:

. Drag the Flow Reference component from *Mule Palette* to the point in your flow where you want to create the call to the referenced flow:
+
image:mruntime-flow-ref-create.png[flow-ref-create]
. Open the Flow Reference properties and specify the flow you want to reference in *Flow name*:
+
image:mruntime-flow-ref-config.png[flow-ref-config]

== Enrich Content with a Flow Reference Target Variable

> You can store the result of a Flow Reference call in a target variable by configuring the following fields in your Flow Reference properties:

Reviewer: Should we add more about WHY the audience wants or needs to do this?

* *Target*: Name of the variable that the Flow Reference creates to store the result of the processed payload.

> * *Target Value*: An expression that evaluates against the operation's output, the result of this evaluation is stored in the defined target variable.

Reviewer: ```suggestion
* *Target Value*: An expression that evaluates against the operation's output. The result of this evaluation is stored in the defined target variable.
```

> === Considerations About Target Variable

Reviewer: Instead of "Considerations," how about "Implications of Using a Target Variable"? If not...:
```suggestion
=== Considerations When Using a Target Variable
```

> Setting the `target` variable in a Flow Reference component causes the message to remain unchanged. This means that any modification to the payload or to the variables that occur in the referenced flow will revert once the referenced flow finishes executing its processors, returning the payload and variables to its original values.

Reviewer: ```suggestion
Setting the `target` variable in a Flow Reference component causes the original message to remain unchanged. This means that any modification to the payload or to the variables that occur in the referenced flow revert after the referenced flow finishes executing its processors, returning the payload and variables to their original values.
```

> On the other hand, if you do not define a `target` variable, any modification to the payload or to the variables that occurs in the referenced flow will persist after the referenced flow finishes its execution, changing the actual value of the payload and variables.

Reviewer: Try to avoid idioms like "on the other hand," to ease translation. How about "alternatively"?
```suggestion
Alternatively, if you do not define a `target` variable, any modification that occurs to the payload or to the variables in the referenced flow persist after the referenced flow finishes its execution, changing the original value of the payload or variables.
```

> === XML Example

Reviewer: Is this more precise?
```suggestion
=== XML Example of Target Variable Use
```

> The following XML code represents the project configured in the previous steps, and also defines target variable `flowReferenceVar` which stores the result of the secondary flow called by the Flow Reference component. This variable is then used from the main flow by logging `vars.flowReferenceVar`:

Reviewer: Because "the previous steps" appear to be rather removed from this text, can we reword by internal xref?
```suggestion
The following XML code represents the project configured in << title >>, adding the target variable `flowReferenceVar,` which stores the result of the secondary flow called by the Flow Reference component. This variable is then used in the main flow by logging `vars.flowReferenceVar`:
```
